crates/sui-tool/src/lib.rs

```diff
@@ -618,6 +620,103 @@ pub async fn restore_from_db_checkpoint(
+    local_store_dir: PathBuf,
+    genesis: Genesis,
+    concurrency: Option<NonZeroUsize>,
+    disable_verification: Option<bool>,
+) -> anyhow::Result<()> {
+    // first ensure that target db is empty
+    if !AuthorityPerpetualTables::open(&parent_db_path.join("store").join("perpetual"), None)
+        .database_is_empty()
+        .expect("Database read should not fail at init.")
+    {
+        panic!("Database is not empty. Will not perform snapshot restoration");
+    }
+
+    let genesis_committee = genesis.committee()?;
+
+    let perpetual_staging = Arc::new(AuthorityPerpetualTables::open(
+        &parent_db_path.join("store_staging").join("perpetual"),
+        None,
+    ));
+
+    let committee_store = Arc::new(CommitteeStore::new(
+        parent_db_path.join("epochs"),
+        &genesis_committee,
+        None,
+    ));
+
+    let store = AuthorityStore::open(
+        perpetual_staging.clone(),
+        &genesis,
+        &committee_store,
+        usize::MAX,
+        false,
+        &Registry::default(),
+    )
+    .await?;
+
+    let checkpoint_store = Arc::new(CheckpointStore::open_tables_read_write(
+        parent_db_path.join("checkpoints"),
+        MetricConf::default(),
+        None,
+        None,
+    ));
+    // Only insert the genesis checkpoint if the DB is empty and doesn't have it already
+    if checkpoint_store
+        .get_checkpoint_by_digest(genesis.checkpoint().digest())
+        .unwrap()
+        .is_none()
+    {
+        checkpoint_store.insert_checkpoint_contents(genesis.checkpoint_contents().clone())?;
+        checkpoint_store.insert_verified_checkpoint(&genesis.checkpoint())?;
+        checkpoint_store.update_highest_synced_checkpoint(&genesis.checkpoint())?;
+    }
+
+    let state_sync_store = RocksDbStore::new(store, committee_store, checkpoint_store);
+
+    let archive_reader_config = ArchiveReaderConfig {
+        remote_store_config: archive_object_store_config,
+        use_for_pruning_watermark: false,
+        download_concurrency: concurrency.unwrap_or(NonZeroUsize::new(3).unwrap()),
+    };
+    let metrics = ArchiveReaderMetrics::new(&Registry::default());
+    let archive_reader = Arc::new(ArchiveReader::new(archive_reader_config, &metrics)?);
+    archive_reader.sync_manifest_once().await?;
+
+    let restorer = SnapshotRestorer::new(
+        archive_reader,
+        parent_db_path.clone(),
+        epoch,
+        snapshot_object_store_config,
+        local_store_dir,
+        concurrency,
+    )
+    .await?;
+    info!("Restoring from snapshot");
+
+    restorer
+        .run(
+            state_sync_store.clone(),
+            perpetual_staging,
+            disable_verification.unwrap_or(false),
+        )
+        .await
+        .expect("Snapshot restoration failed");
+
+    // Hardlink staging tables to perpetual store path so that on restart, the
+    // data will be available in the expected authority store directory.
+    Ok(hard_link(
```

Comment:
If we used `/opt/sui/db/authority_db/staging` path, after successful restore we could just `mv /opt/sui/db/authority_db/staging /opt/sui/db/authority_db/live` which is an atomic operation

Reply:
I was unsure whether something like that would work, given that that is a syscall, and wouldn't necessarily change any kind of path bindings within the rocks db store object. Sounds good will give this a try